### PR TITLE
Documenting AWS Batch set up

### PR DESCRIPTION
**Pull request by lboyarsky**
Ticket: [document the process for building tile builder](https://app.asana.com/0/810933294009540/1125371268126433/f)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -76,3 +76,38 @@ $ docker run --tty \
...
+### Publishing Tiles through AWS Batch
+
+To generate tiles for production usage we use [AWS Batch tool](https://aws.amazon.com/batch/), which allows to execute multiple similar jobs in parallel. In order to use AWS Batch, the first prerequisite is to push the docker image into AWS ECR Repository. Tile server image is already there so this step is only required in case if you have done any changes to the image (that is, changes to dockerfiles itself or to the files used by it). In order to push the image to ECR Repository, execute following commands: 
+
+* `$(aws ecr get-login --no-include-email --region us-east-1)`: login to ECR
+* `docker build -t tile-server .`: build tile server docker image
+* `docker tag tile-server:latest 434035161053.dkr.ecr.us-east-1.amazonaws.com/tile-server:latest`: tag the image with the version called "latest"
+* `docker push 434035161053.dkr.ecr.us-east-1.amazonaws.com/tile-server:latest`: push (upload) docker image to the repository
+
+These steps should normally be set up to run in CI system, but at the time of writing this we don't have one, so they need to be executed manually.
```
Comment (paulswartz) [inline README.md:89]: do we have a ticket to do this?

File: README.md [hunk 2]
```diff
@@ -76,3 +76,38 @@ $ docker run --tty \
...
+
+In order to launch AWS Batch jobs, do the following:
+* Login into AWS Console and navigate to AWS Batch: https://console.aws.amazon.com/batch/
+* Go to "Job Definitions" section, click on "tile-generation-dev" item and then on the radio-button corresponding to the latest revision. 
+* Click Actions -> Submit job
+* On the next screen do the following:
+  * Specify job name, which describes current job the best (no special requirements for the name)
+  * Specify `tile-generation-dev-queue` as the job queue name. 
+  * **Important**: specify **Array** as job type. 
+  * Specify number of jobs to be run as parallel in Array Size field. If you don't know how many you need, put `8`. It took 5-6 hours for 8 jobs to generate the tiles for entire service area up to maximum zoom level (18 during previous tests.
```
Comment (paulswartz) [inline README.md]: What's the reason for 8? vs some other larger or smaller #? Also the line is missing a closing `)`.

File: README.md [hunk 3]
```diff
@@ -76,3 +76,38 @@ $ docker run --tty \
...
+  * Navigate to **Environment variables** section and make sure that we have both mandatory variables set to correct values:
+    * `MAPNIK_TILE_S3_BUCKET`: should correspond with the target S3 bucket; `mbta-map-tiles-dev` for dev environment or `mbta-map-tiles` for production environment. If you want to upload files to different bucket, you can specify it here as well, but make sure that batch jobs have permissions to write there. 
+    * **Important**: `BATCH_JOB_COUNT`: should be equal to what you specified in Array Size field. Failure to set the right value here will result in either missing tiles or in overal between the jobs (they are going to generate the same tiles).
+  * Click "Submit Job" and wait for jobs to run. You can also check Cloud Watch logs, while the jobs are running. Links to those logs are available from the Jobs page. 
+
+These steps should also normally be set up to run in CI system, but at the time of writing this we don't have one, so they need to be executed manually.
```
Comment (paulswartz) [inline README.md:113]: is there a ticket to do this? (possibly part of the same CI as the Docker file build)

File: README.md (hunk 1 above)
Comment (lboyarsky) [reply README.md:89]: @ianwestcott has it: https://app.asana.com/0/1113179098808463/1116455779572731/f

File: README.md (hunk 2 above)
Comment (lboyarsky) [reply README.md]: `8` seems to be balanced in terms of running time vs costs vs overhead. I haven't done extensive testing/measures though. I've added closing `)` in previous commit.

File: README.md (hunk 3 above)
Comment (lboyarsky) [reply README.md:113]: I think it's the same @ianwestcott's task - https://app.asana.com/0/1113179098808463/1116455779572731/f

**Review by paulswartz — APPROVED** [review]
At some point we should document how to clean up the old batch jobs so they're not hanging around the ECS listing.

**Comment by lboyarsky** [comment]
@paulswartz I think they are cleaned up automatically after some time. At least I don't see any older jobs now, only the job from yesterday, and I haven't ever cleaned them.

File: README.md (hunk 3 above)
Comment (ianwestcott) [reply README.md:113]: It's actually this task, which was recently completed as Won't Do: https://app.asana.com/0/947623132087374/1117279835834479 Feel free to reopen it and advocate for it if you want to see it done.

In the meantime, we could write a shell script to encapsulate the logic in anticipation of it being run by automation.

File: README.md [hunk 4]
```diff
@@ -76,3 +76,38 @@ $ docker run --tty \
...
+* On the next screen do the following:
+  * Specify job name, which describes current job the best (no special requirements for the name)
+  * Specify `tile-generation-dev-queue` as the job queue name. 
+  * **Important**: specify **Array** as job type. 
+  * Specify number of jobs to be run as parallel in Array Size field. If you don't know how many you need, put `8`. It took 5-6 hours for 8 jobs to generate the tiles for entire service area up to maximum zoom level (18) during previous tests. 
+  * Navigate to **Environment variables** section and make sure that we have both mandatory variables set to correct values:
+    * `MAPNIK_TILE_S3_BUCKET`: should correspond with the target S3 bucket; `mbta-map-tiles-dev` for dev environment or `mbta-map-tiles` for production environment. If you want to upload files to different bucket, you can specify it here as well, but make sure that batch jobs have permissions to write there. 
+    * **Important**: `BATCH_JOB_COUNT`: should be equal to what you specified in Array Size field. Failure to set the right value here will result in either missing tiles or in overal between the jobs (they are going to generate the same tiles).
```
Comment (ianwestcott) [inline README.md]: `overal` → `overlap`

File: README.md (hunk 3 above)
Comment (lboyarsky) [reply README.md:113]: I still suggest that it's the task I linked - https://app.asana.com/0/947623132087374/1116455779572731/f - its title ("Set up periodic task for tile-server image generation/publishing") refers to the that same activity as I'm describing here. 

I don't think there any doubts about whether this is required, or are there, @gshkolnik? If we don't do it our maps are going to become outdated quickly.

**Review by ianwestcott — COMMENTED** [review]
Found one typo. Will run through the process today or tomorrow and note if I hit any snags.

File: README.md (hunk 3 above)
Comment (ianwestcott) [reply README.md:113]: Sorry, I might have replied to the wrong comment. But I gather there are two distinct tasks at issue regarding automation:

1. [Building and publishing the container via CI](https://app.asana.com/0/947623132087374/1117279835834479)
1. [Running tile generation periodically](https://app.asana.com/0/947623132087374/1116455779572731/f)

The latter is in my backlog. The former was marked "Won't Do" but IMO we can add a script to the repo in lieu of creating a Semaphore project.

File: README.md (hunk 4 above)
Comment (lboyarsky) [reply README.md]: Fixed

**Review by gshkolnik — CHANGES_REQUESTED** [review]
Lev, the document looks great. However, before you merge this, I would like to ask @ianwestcott to go through these steps to make sure he is actually able to follow them. @ianwestcott  -  could you please do this and then approve this PR?

File: README.md [hunk 5]
```diff
@@ -76,3 +76,38 @@ $ docker run --tty \
...
+
+The idea behind running the jobs in AWS batch is that we split the _service area_ into horizontal (East-West) stripes and generate tiles for each of those map stripes independently. Since there is no overlap, the execution can be parallelized and we can have as many parallel jobs as we see necessary. 
+
+Important thing to note is that the service area only make sense if we have the map data for it. Currently we have map data for MA, RI and NH, so if you need service area to cover any other territories, you need to add map data first ([example PR](https://github.com/mbta/tile-server/pull/12/files)).
+
+In order to launch AWS Batch jobs, do the following:
+* Login into AWS Console and navigate to AWS Batch: https://console.aws.amazon.com/batch/
+* Go to "Job Definitions" section, click on "tile-generation-dev" item and then on the radio-button corresponding to the latest revision.
```
Comment (ianwestcott) [inline README.md]: As of last night there are now two different job definitions and job queues: one for dev and one for prod. Can you update these steps to reflect that the user should now choose one or the other?

File: README.md [hunk 6]
```diff
@@ -76,3 +76,38 @@ $ docker run --tty \
...
+* Go to "Job Definitions" section, click on "tile-generation-dev" item and then on the radio-button corresponding to the latest revision. 
+* Click Actions -> Submit job
+* On the next screen do the following:
+  * Specify job name, which describes current job the best (no special requirements for the name)
+  * Specify `tile-generation-dev-queue` as the job queue name. 
+  * **Important**: specify **Array** as job type. 
+  * Specify number of jobs to be run as parallel in Array Size field. If you don't know how many you need, put `8`. It took 5-6 hours for 8 jobs to generate the tiles for entire service area up to maximum zoom level (18) during previous tests. 
+  * Navigate to **Environment variables** section and make sure that we have both mandatory variables set to correct values:
+    * `MAPNIK_TILE_S3_BUCKET`: should correspond with the target S3 bucket; `mbta-map-tiles-dev` for dev environment or `mbta-map-tiles` for production environment. If you want to upload files to different bucket, you can specify it here as well, but make sure that batch jobs have permissions to write there.
```
Comment (ianwestcott) [inline README.md]: Now that there are separate jobs for dev and prod (and I updated the IAM roles to restrict each to the respective environment's bucket) it's no longer necessary to modify this field.

File: README.md [hunk 7]
```diff
@@ -76,3 +76,38 @@ $ docker run --tty \
...
+  * **Important**: specify **Array** as job type. 
+  * Specify number of jobs to be run as parallel in Array Size field. If you don't know how many you need, put `8`. It took 5-6 hours for 8 jobs to generate the tiles for entire service area up to maximum zoom level (18) during previous tests. 
+  * Navigate to **Environment variables** section and make sure that we have both mandatory variables set to correct values:
+    * `MAPNIK_TILE_S3_BUCKET`: should correspond with the target S3 bucket; `mbta-map-tiles-dev` for dev environment or `mbta-map-tiles` for production environment. If you want to upload files to different bucket, you can specify it here as well, but make sure that batch jobs have permissions to write there. 
+    * **Important**: `BATCH_JOB_COUNT`: should be equal to what you specified in Array Size field. Failure to set the right value here will result in either missing tiles or in overlap between the jobs (they are going to generate the same tiles).
+  * Click "Submit Job" and wait for jobs to run. You can also check Cloud Watch logs, while the jobs are running. Links to those logs are available from the Jobs page.
```
Comment (ianwestcott) [inline README.md]: Is it worth mentioning the [SNS topic](https://console.aws.amazon.com/sns/v3/home?region=us-east-1#/topic/arn:aws:sns:us-east-1:434035161053:batch-tile-generation-job-notify) here? I'm not sure if everyone will have the permissions to subscribe to it, but they could request to be added. Or we could set up a mailing list, as proposed in https://app.asana.com/0/947623132087374/1118451804049276/f.

**Review by ianwestcott — COMMENTED** [review]
Some comments, mostly based on changes I've made in the last day. Overall the process works as described.

File: README.md (hunk 7 above)
Comment (ianwestcott) [reply README.md]: Update: I added a Slack integration to the #tile-server channel.

File: README.md (hunk 5 above)
Comment (lboyarsky) [reply README.md]: Done

File: README.md (hunk 6 above)
Comment (lboyarsky) [reply README.md]: Removed.

File: README.md (hunk 7 above)
Comment (lboyarsky) [reply README.md]: Added

**Comment by lboyarsky** [comment]
@ianwestcott is it good to merge?

File: README.md [hunk 8]
```diff
@@ -98,16 +98,16 @@ Important thing to note is that the service area only make sense if we have the
...
-  * Specify number of jobs to be run as parallel in Array Size field. If you don't know how many you need, put `8`. It took 5-6 hours for 8 jobs to generate the tiles for entire service area up to maximum zoom level (18) during previous tests. 
-  * Navigate to **Environment variables** section and make sure that we have both mandatory variables set to correct values:
-    * `MAPNIK_TILE_S3_BUCKET`: should correspond with the target S3 bucket; `mbta-map-tiles-dev` for dev environment or `mbta-map-tiles` for production environment. If you want to upload files to different bucket, you can specify it here as well, but make sure that batch jobs have permissions to write there. 
-    * **Important**: `BATCH_JOB_COUNT`: should be equal to what you specified in Array Size field. Failure to set the right value here will result in either missing tiles or in overlap between the jobs (they are going to generate the same tiles).
-  * Click "Submit Job" and wait for jobs to run. You can also check Cloud Watch logs, while the jobs are running. Links to those logs are available from the Jobs page. 
+  * Specify `tile-generation-dev-queue` as the job queue name
```
Comment (ianwestcott) [inline README.md]: Should be something like "specify `tile-generation-dev-queue` for dev environment or `tile-generation-prod-queue` for production environment"

**Review by ianwestcott — APPROVED** [review]
One more dev/prod node, otherwise good to go. Thanks!

File: README.md (hunk 8 above)
Comment (lboyarsky) [reply README.md]: Fixed

**Pull request merged**